From f7eb0a072858cb5a1e815957499f62d0fbc23120 Mon Sep 17 00:00:00 2001
From: Nicolas Le Bayon <nicolas.le.bayon@foss.st.com>
Date: Thu, 3 Apr 2025 13:21:50 +0200
Subject: [PATCH 827/841] fix(st-crypto): parse RNG DT before interacting with
 the IP

STM32MP21 HWBUS mechanism is automatically started when RNG2 clock is
enabled, unless IP reset is asserted (done just after clock enable())
or CONDRST is activated in RNG_CR register (done in stm32_rng_enable
function). In the interval, software must ensure that time consumption
is not long enough to provide words to BSEC in poor conditions (reset
values of RNG2 registers).

parse_dt_optional_config() doesn't interact with the RNG IP and only
treats device tree options within RNG node. Time comsumption is quite
important. Move the call of the function before enabling the clock to
ensure HWBUS word generation is not possible. It can be applied for
all platforms.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 797ba0555fbbb19ce39c0f22640ce0ea9431a52f fix(st-crypto): parse RNG DT before interacting with the IP]
Signed-off-by: Nicolas Le Bayon <nicolas.le.bayon@foss.st.com>
Change-Id: I5120f0f4119697f443d17e164733311bbf9c8390
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/454480
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Tested-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
Reviewed-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 drivers/st/crypto/stm32_rng.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/st/crypto/stm32_rng.c b/drivers/st/crypto/stm32_rng.c
index 711839ffa..a99354f58 100644
--- a/drivers/st/crypto/stm32_rng.c
+++ b/drivers/st/crypto/stm32_rng.c
@@ -374,6 +374,8 @@ int stm32_rng_init(void)
 			panic();
 		}
 
+		parse_dt_optional_config(fdt, node);
+
 		stm32_rng.clock = (unsigned long)dt_rng.clock;
 		clk_enable(stm32_rng.clock);
 
@@ -397,8 +399,6 @@ int stm32_rng_init(void)
 			}
 		}
 
-		parse_dt_optional_config(fdt, node);
-
 		ret = stm32_rng_enable();
 		if (ret != 0) {
 			ERROR("Failed to enable rng@%x\n", dt_rng.base);
-- 
2.47.2

